diff --git a/buildfiles/libraw.pro b/buildfiles/libraw.pro
index 515a5e1..3e5c5ca 100644
--- a/buildfiles/libraw.pro
+++ b/buildfiles/libraw.pro
@@ -1,17 +1,19 @@
 TEMPLATE=lib
 TARGET=libraw
-INCLUDEPATH+=../
+INCLUDEPATH+=$$PWD/../
 include (libraw-common-lib.pro)
 # JPEG section (libjpeg is linked later)
 DEFINES+=USE_JPEG
+DEFINES+=NO_LCMS
 
 ## RawSpeed section
 DEFINES+=USE_RAWSPEED
-INCLUDEPATH+=../../RawSpeed/
-win32:INCLUDEPATH+= $$TJDIR/include
-win32:INCLUDEPATH+=d:/Qt/local/include d:/Qt/local/include/libxml2
-win32:LIBS+=-L$$TJDIR/lib$$TJSUFF/$$SUFF  -ljpeg 
-win32:LIBS+=-lrawspeed -Ld:/Qt/local/lib/$$SUFF -llibxml2
+INCLUDEPATH+=$$PWD/../../rawspeed/
+INCLUDEPATH+=$$PWD/../../rawspeed/include
+win32:INCLUDEPATH *= $$PWD/../../third-party/libxml2/include $$PWD/../../third-party/libjpeg-turbo/include
+win32:LIBS+=-lws2_32
+win32:LIBS+=-L$$PWD/../../third-party/libjpeg-turbo/lib -ljpeg
+win32:LIBS+=-L$$OUT_PWD/../../libs -lrawspeed -L$$PWD/../../third-party/libxml2/lib -llibxml2
 
 HEADERS=../libraw/libraw.h \
 	 ../libraw/libraw_alloc.h \
@@ -24,7 +26,7 @@ HEADERS=../libraw/libraw.h \
 	../internal/var_defines.h \
 	../internal/libraw_internal_funcs.h
 
-win32: {
+win32-msvc: {
 PREPROCESS_FILES=../dcraw/dcraw.c
 preprocess.name=dcraw.c preprocess
 preprocess.input=PREPROCESS_FILES
@@ -50,3 +52,8 @@ SOURCES+= \
 	../src/libraw_datastream.cpp \
 	../src/libraw_c_api.cpp
 
+DESTDIR = ../../libs
+
+## Installation
+target.path = $$[QT_INSTALL_LIBS]
+INSTALLS = target
